src/analysis/convex/extreme.lean

```diff
@@ -0,0 +1,286 @@
+/-
+Copyright (c) 2021 Yaël Dillies, Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import linear_algebra.affine_space.independent
+import algebra.big_operators.basic
+import analysis.convex.topology
+
+/-!
+# Extreme sets
+This file defines extreme sets and extreme points for set in a real normed space.
+## References
+See chapter 8 of [Convexity][simon2011]
+TODO:
+- add exposed sets to this file.
+- define convex independence and prove lemmas related to extreme points.
+-/
+
+open_locale classical affine
+open set
+
+variables {E : Type*} [add_comm_group E] [vector_space ℝ E] {x : E} {A B C : set E}
+
+/--
+A set B is extreme to a set A if B ⊆ A and all points of B only belong to (intrinsic) interiors of
+segments whose ends are in B.
+-/
+def is_extreme_set (A B : set E) :
+  Prop :=
+B ⊆ A ∧ ∀ x₁ x₂ ∈ A, ∀ x ∈ B, x ∈ segment x₁ x₂ → x₁ ≠ x → x₂ ≠ x → x₁ ∈ B ∧ x₂ ∈ B
+
+lemma is_extreme_set.refl :
+  reflexive (is_extreme_set : set E → set E → Prop) :=
+λ A, ⟨subset.refl _, λ x₁ x₂ hx₁A hx₂A x hxA hx hxx₁ hxx₂, ⟨hx₁A, hx₂A⟩⟩
+
+lemma is_extreme_set.trans :
+  transitive (is_extreme_set : set E → set E → Prop) :=
+begin
+  rintro A B C ⟨hBA, hAB⟩ ⟨hCB, hBC⟩,
+  use subset.trans hCB hBA,
+  rintro x₁ x₂ hx₁A hx₂A x hxC hx hxx₁ hxx₂,
+  obtain ⟨hx₁B, hx₂B⟩ := hAB x₁ x₂ hx₁A hx₂A x (hCB hxC) hx hxx₁ hxx₂,
+  exact hBC x₁ x₂ hx₁B hx₂B x hxC hx hxx₁ hxx₂,
+end
+
+lemma is_extreme_set.antisymm :
+  anti_symmetric (is_extreme_set : set E → set E → Prop) :=
+λ A B hAB hBA, subset.antisymm hBA.1 hAB.1
+
+instance : is_partial_order (set E) is_extreme_set :=
+{ refl := is_extreme_set.refl,
+  trans := is_extreme_set.trans,
+  antisymm := is_extreme_set.antisymm }
+
+lemma convex_diff_of_extreme (hA : convex A) (hAB : is_extreme_set A B) :
+  convex (A \ B) :=
+begin
+  rw convex_iff_segment_subset,
+  rintro x₁ x₂ ⟨hx₁A, hx₁B⟩ ⟨hx₂A, hx₂B⟩ x hx,
+  refine ⟨hA.segment_subset hx₁A hx₂A hx, λ hxB, hx₁B (hAB.2 x₁ x₂ hx₁A hx₂A x hxB hx _ _).1⟩,
+  { rintro rfl,
+    exact hx₁B hxB },
+  { rintro rfl,
+    exact hx₂B hxB }
+end
+
+/--
+A point x is an extreme point of a set A if x belongs to no (intrinsic) interior of a segment with
+ends in A.
+-/
+def set.extreme_points (A : set E) :
+  set E :=
+{x ∈ A | ∀ (x₁ x₂ ∈ A), x ∈ segment x₁ x₂ → x₁ = x ∨ x₂ = x}
+
+lemma extreme_point_def :
+  x ∈ A.extreme_points ↔ x ∈ A ∧ ∀ (x₁ x₂ ∈ A), x ∈ segment x₁ x₂ → x₁ = x ∨ x₂ = x :=
+by refl
+
+/--
+x is an extreme point to A iff {x} is an extreme set of A.
+-/
+lemma extreme_point_iff_extreme_singleton :
```

Comment:
I think some of the names could be adjusted slightly to match the statement a bit better: eg this could be `mem_extreme_points_iff_is_extreme_set_singleton`? Although maybe others will have better suggestions. Essentially if the lemma is about `is_extreme_set` or something, the name should talk about "is_extreme_set" rather than "extreme"